File: src/decoders/timetable.ts
import type { SessionHandle, Timetable, TimetableClass } from "~/models";
import { decodeTimetableClassActivity } from "./timetable-class-activity";
import { decodeTimetableClassDetention } from "./timetable-class-detention";
import { decodeTimetableClassLesson } from "./timetable-class-lesson";

const isTimetableClassActivity = (item: any): boolean => {
  return "estSortiePedagogique" in item && item.estSortiePedagogique;
};

const isTimetableClassDetention = (item: any): boolean => {
  return "estRetenue" in item && typeof item.estRetenue !== "undefined";
};

export const decodeTimetable = (timetable: any, session: SessionHandle): Timetable => {
  return {
    absences: timetable.absences,
    withCanceledClasses: timetable.avecCoursAnnule ?? true,

    classes: (timetable.ListeCours ?? []).map((item: any) => {
      if (isTimetableClassActivity(item)) {
        return decodeTimetableClassActivity(item, session);
      }

      if (isTimetableClassDetention(item)) {
        return decodeTimetableClassDetention(item, session);
      }

      return decodeTimetableClassLesson(item, session);
    })
      .sort((a: TimetableClass, b: TimetableClass) => (
        a.startDate.getTime() - b.startDate.getTime()
      ))
  };
};

File: library/src/commonMain/kotlin/decoders/timetable.kt
package decoders

import kotlinx.serialization.json.*
import models.SessionHandle
import models.Timetable

val isTimetableClassActivity: (item: JsonObject) -> Boolean = {item -> item.containsKey("estSortiePedagogique") && item["estSortiePedagogique"]!!.jsonPrimitive.boolean}
val isTimetableClassDetention: (item: JsonObject) -> Boolean = {item -> item.containsKey("estRetenue") && item["estRetenue"]!!.jsonPrimitive.content != "undefined"}

fun decodeTimetable (timetable: JsonObject, session: SessionHandle) = Timetable(
    absences = timetable["absences"]!!.jsonObject,
    withCanceledClasses = timetable["avecCoursAnnule"]?.jsonPrimitive?.boolean ?: true,
    classes = timetable["ListeCours"]!!.jsonArray.map { item ->
        val itemObj = item.jsonObject

        val decoder: (item: JsonObject) -> Any = if (isTimetableClassActivity(itemObj)) {
            {decodeTimetableClassActivity(itemObj, session.information)}
        } else if (isTimetableClassDetention(itemObj)) {
            { decodeTimetableClassDetention(itemObj) }
        } else {
            { decodeTimetableClassLesson(itemObj) }
        }

        decodeTimetableClass(itemObj, session, decoder)
    }
)